src/bin/elasticsearch_exporter.rs

```diff
@@ -0,0 +1,359 @@
+#![feature(str_split_once)]
+
+#[macro_use]
+extern crate lazy_static;
+#[macro_use]
+extern crate prometheus;
+#[macro_use]
+extern crate log;
+
+use clap::Clap;
+use hyper::server::conn::AddrStream;
+use hyper::{
+    header::CONTENT_TYPE,
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use prometheus::{Encoder, HistogramVec, TextEncoder, TEXT_FORMAT};
+use std::convert::Infallible;
+use std::env;
+use std::error::Error as StdError;
+use std::fmt;
+use std::net::SocketAddr;
+use std::panic;
+use std::str::FromStr;
+use std::time::Duration;
+use tokio::signal;
+use tokio::sync::oneshot::{self, Receiver, Sender};
+use url::Url;
+
+use elasticsearch_exporter::{
+    CollectionLabels, Exporter, ExporterOptions, ExporterPollIntervals, Labels,
+};
+
+lazy_static! {
+    static ref HTTP_REQ_HISTOGRAM: HistogramVec = register_histogram_vec!(
+        "http_request_duration_seconds",
+        "The HTTP request latencies in seconds.",
+        &["handler"]
+    )
+    .expect("valid histogram vec metric");
+}
+
+fn build_response(status: StatusCode, body: Body) -> Response<Body> {
+    Response::builder()
+        .status(status)
+        .header(CONTENT_TYPE, TEXT_FORMAT)
+        .body(body)
+        .expect("valid Response built")
+}
+
+async fn serve_req(req: Request<Body>, options: String) -> Result<Response<Body>, Infallible> {
+    let path = req.uri().path();
+
+    let timer = HTTP_REQ_HISTOGRAM.with_label_values(&[path]).start_timer();
+
+    let response = match path {
+        "/health" | "/healthy" | "/healthz" => build_response(StatusCode::OK, Body::from("Ok")),
+        "/" => build_response(StatusCode::OK, Body::from(options.to_string())),
+
+        "/metrics" => {
+            let encoder = TextEncoder::new();
+
+            let mut buffer = vec![];
+            match encoder.encode(&prometheus::gather(), &mut buffer) {
+                Ok(_) => build_response(StatusCode::OK, Body::from(buffer)),
+                Err(e) => {
+                    error!("prometheus encoder err {}", e);
+
+                    build_response(StatusCode::INTERNAL_SERVER_ERROR, Body::empty())
```

Comment:
I'd probably extract this into a static value.

Reply:
IMO it is easier to leave this as is, because now it takes less code :see_no_evil:

Reply:
I was thinking more about resource usage. I've expected that if we put that into static, we wouldn't be allocating new response  every single time (although for internal server error it wouldn't be too often anyway)